Stripe/STPCustomer.m

```diff
@@ -85,43 +83,61 @@ + (instancetype)decodedObjectFromAPIResponse:(NSDictionary *)response {
         shipping.country = shippingDict[@"address"][@"country"];
         customer.shippingAddress = shipping;
     }
-    NSMutableArray *sources = [NSMutableArray array];
-    if ([dict[@"sources"] isKindOfClass:[NSDictionary class]] && [dict[@"sources"][@"data"] isKindOfClass:[NSArray class]]) {
-        for (id contents in dict[@"sources"][@"data"]) {
-            if ([contents isKindOfClass:[NSDictionary class]]) {
-                if ([contents[@"object"] isEqualToString:@"card"]) {
-                    STPCard *card = [STPCard decodedObjectFromAPIResponse:contents];
-                    // ignore apple pay cards from the response
-                    if (card && !card.isApplePayCard) {
-                        [sources addObject:card];
-                        if (defaultSourceId && [card.stripeID isEqualToString:defaultSourceId]) {
-                            customer.defaultSource = card;
-                        }
+    [customer updateSourcesWithResponse:dict filteringApplePay:YES];
+    customer.allResponseFields = dict;
+    return customer;
+}
+
+- (void)updateSourcesWithResponse:(NSDictionary *)response
+                filteringApplePay:(BOOL)filterApplePay {
+    NSArray *data;
+    if ([response[@"sources"] isKindOfClass:[NSDictionary class]]) {
+        data = response[@"sources"][@"data"];
+    }
+    if (![data isKindOfClass:[NSArray class]]) {
+        return;
```

Comment:
When called from `decodedObjectFromAPIResponse:`, this will return a customer with `sources = nil` (if `sources.data` is not an `NSArray`), but the property declaration for `sources` is `nonnull`

I wasn't (and still am not) sure how we want to handle a malformed `response`. I like that we're not crashing! :)

I also thought I liked the idea that you're not resetting `sources` and `defaultSources` if this is not the initialization pass. However, this is private, and it looks like you're always just using the cached response, so maybe it'd be easier to just reset to empty array & nil?

Another option would be to set sources to the empty array in `decodedObjectFromAPIResponse:` before calling this method. I think I like this option slightly better.

Reply:
ah good catch, I agree that the second option is better – I'll update. Working on some related guards against malformed responses in my next PR :)

Reply:
updated in ffb5ff6